SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp

```diff
@@ -329,29 +330,39 @@ void ASpatialDebugger::DrawDebugLocalPlayer(UCanvas* Canvas)
 	}
 }
 
-const FColor& ASpatialDebugger::GetVirtualWorkerColor(const Worker_EntityId EntityId) const
+FColor ASpatialDebugger::GetVirtualWorkerColor(const Worker_EntityId EntityId) const
 {
 	check(NetDriver != nullptr && !NetDriver->IsServer());
-
 	const AuthorityIntent* AuthorityIntentComponent = NetDriver->StaticComponentView->GetComponentData<AuthorityIntent>(EntityId);
-	const int32 VirtualWorkerId = (AuthorityIntentComponent != nullptr) ? AuthorityIntentComponent->VirtualWorkerId : SpatialConstants::INVALID_VIRTUAL_WORKER_ID;
-
-	if (VirtualWorkerId != SpatialConstants::INVALID_VIRTUAL_WORKER_ID &&
-		VirtualWorkerId < ServerTintColors.Num())
-	{
-		return ServerTintColors[VirtualWorkerId];
-	}
-	else
-	{
+	check(AuthorityIntentComponent != nullptr);
+	const int32 VirtualWorkerId = AuthorityIntentComponent->VirtualWorkerId;
+	const FString* PhysicalWorkerId = NetDriver->VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(VirtualWorkerId);
+	if (PhysicalWorkerId == nullptr) {
+		// This can happen if the client hasn't yet received the VirtualWorkerTranslator mapping
 		return InvalidServerTintColor;
 	}
+	return SpatialGDK::GetColourForWorkerId(*PhysicalWorkerId);
 }
 
-// TODO: Implement once this functionality is available https://improbableio.atlassian.net/browse/UNR-2362.
-const FColor& ASpatialDebugger::GetServerWorkerColor(const Worker_EntityId EntityId) const
+FColor ASpatialDebugger::GetServerWorkerColor(const Worker_EntityId EntityId) const
 {
 	check(NetDriver != nullptr && !NetDriver->IsServer());
-	return InvalidServerTintColor;
+
+	const SpatialGDK::EntityAcl* AclData = NetDriver->StaticComponentView->GetComponentData<SpatialGDK::EntityAcl>(EntityId);
+	const WorkerRequirementSet* WriteAcl = AclData->ComponentWriteAcl.Find(SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID);
```

Comment:
Is looking at the entity ACL the correct way to report which worker has authority? I thought there would be a user-defined component which is updated by the authoritative server when it gains authority. I guess it matters what type of information is most useful for the user of the debugger - in the description for the flag which enables this functionality the description is "Show server authority for every entity in range" so I assumed this will display the authority only after it has been granted by the runtime and the acked by the server which received it. With the current implementation, this server worker could be in the process of being granted authority.

Reply:
> I thought there would be a user-defined component which is updated by the authoritative server when it gains authority
this is how I'm proposing the per-actor debugger logic works, a new actor component (debug mode only) that contains the relevant virtual worker / authoritative worker name, and then none of this lookup need exist